Misc fixes to the localemod module

asomers opened the pull request:
### What does this PR do?
1) Fixes the error message for locale.get_locale on unsupported platforms.  Previously it would raise a KeyError
2) Skips the localemod tests on FreeBSD, where that module is unsupported
3) Fixes some localemod tests whose pass conditions were too generous.

### What issues does this PR fix or reference?
none

### Previous Behavior
Calling localemod.get_locale on an unsupported platform would raise an unhelpful KeyError

### New Behavior
Calling localemod.get_locale on an unsupported platform would raise a CommandExecutionError, as it should

### Tests written?
No.  The existing localemod tests are now skipped on FreeBSD.  If any new errors pop up in CI, it's probably because I fixed test_gen_locale and test_set_locale.  Previously, those tests weren't actually asserting anything

### Commits signed with GPG?
Yes

asomers committed:
Fix the error message for locale.get_locale on unsupported platforms

Now it will return an "XXX is unsupported!" error, whereas before it
would raise a KeyError.

asomers committed:
"Fix" LocaleModuleTest.{test_gen_locale,test_set_locale}

These tests were passing everywhere because their assertions were
ineffective.  Correctly the assertions results in the tests failing, as
expected, on FreeBSD.

asomers committed:
Skip LocaleModuleTest on FreeBSD

This module does not support FreeBSD

dwoz commented:
@asomers looks like the test failures are related to the change.

asomers commented:
Yeah.  I'll set up a Linux development environment so I can debug those.

s0undt3ch commented on tests/integration/modules/test_localemod.py:
```diff
@@ -46,7 +47,7 @@ def test_gen_locale(self):
         locale = self.run_function('locale.get_locale')
         new_locale = _find_new_locale(locale)
         ret = self.run_function('locale.gen_locale', [new_locale])
-        self.assertTrue(ret)
```
We should still assert against `ret`, we're testing against `gen_locale`.
Probably `get_locale` again using the new locale, and then assert against the error not being present.

asomers replied:
asserting against `ret` is useless.  It always succeeds, even if the command failed.

s0undt3ch replied:
Hmm, that suggests a bad implementation.
Not saying it should be addressed in this PR, at least for now.

asomers committed:
fixup to "Fix the error message for locale.get_locale"

Change the error message in the expectation for
unit.modules.test_localemod.LocalemodTestCase.test_get_locale_with_no_systemd_unknown

asomers commented:
The failure on ci/py3/centos7/proxy was due to bundler, not my changes.

asomers committed:
fixup to "Fix LocaleModuleTest.{test_gen_locale,test_set_locale}"

Check ret the correct way

asomers committed:
fixup

asomers-ax committed:
Check ret the correct way, for real this time

run_function can return different types depending on which command was
executed.  The gen_locale and set_locale commands return booleans on success,
or strings on error.  So the correct way to test their result is to check
that the output "is" True, not that it "equals" True.

asomers committed:
Fix locale.gen_locale on CentOS 8 and probably Fedora

CentOS 8 (but not 7), Fedora 30, and Suse store locales in a different
location than Debian-based distros.  Rather than trying to hardcode this
information into saltstack, simply check both locations.

asomers committed:
Skip LocaleModuleTest.test_set_locale if charmaps are not available

This test needs to generate locales, too.

asomers commented:
I believe I fixed the failures on CentOS and Fedora

asomers committed:
Fixup to "Fix locale.gen_locale on CentOS 8"

It's necessary to change the error message, since Salt now searches
multiple locations for the locales

dwoz commented:
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.
